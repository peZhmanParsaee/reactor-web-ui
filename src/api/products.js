import apiCall from '../utils/api-call';

export const addProduct = (product) =>
  apiCall({
    url: 'v1/product',
    method: 'POST',
    data: product
  });

export const getProducts = () => apiCall({ url: 'v1/product' });

export const getProvinces = () => apiCall({ url: 'v1/province' });
