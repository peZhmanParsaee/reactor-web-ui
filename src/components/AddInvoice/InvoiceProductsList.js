import React from 'react';
import {
  List, ListItem, Select, MenuItem, TextField
} from '@material-ui/core';

// @material-ui/icons
import Icon from '@material-ui/core/Icon';
import IconButton from '@material-ui/core/IconButton';

import InvoiceProductsListItem from './InvoiceProductsListItem';
import { separateDigits } from '../../helpers/numberHelpers';

export default InvoiceProductsList = (props) => {
  const onAddNewProductToInvoice = () => {

  };

  return (
    <List>
      <ListItem className={props.classes.addFactorProductsListItemHead} key="iplh_1">
        <List className={props.classes.addFactorProductsNestedList}>
          <ListItem style={{width:'30%'}}
            className={props.classes.addFactorProductsNestedListItemHead}
            key={`iplh_1_1`}
          >نام محصول</ListItem>
          <ListItem style={{width:'20%'}}
            className={props.classes.addFactorProductsNestedListItemHead}
            key={`iplh_1_2`}
          >تعداد</ListItem>
          <ListItem style={{width:'20%'}}
            className={props.classes.addFactorProductsNestedListItemHead}
            key={`iplh_1_3`}>قیمت واحد</ListItem>
          <ListItem style={{width:'20%'}}
            className={props.classes.addFactorProductsNestedListItemHead}
            key={`iplh_1_4`}>قیمت کل</ListItem>
          <ListItem style={{width:'10%'}}
            key={`iplh_1_5`}></ListItem>
        </List>
      </ListItem>
      {
        props.products.map((product) => {
          return <InvoiceProductsListItem 
            key={`ipli1_${product.id}`}
            itemKey={`ipli_${product.id}_i`}
            onAddedProductCountChange={props.onAddedProductCountChange}
            product={{
              id: product.id,
              name: product.name,
              unitPrice: product.unitPrice,                              
              count: product.count,
              totalPrice: product.totalPrice
            }}
          />
        })
      }
      
      <ListItem className={props.classes.addFactorProductsListItem} key={`ipli_add`}>
        <List className={props.classes.addFactorProductsNestedList}>
          <ListItem style={{width:'30%'}}
            className={props.classes.addFactorProductsNestedListItem}
            key={`ipli_add_1`}
          >
            <IconButton onClick={props.onOpenAddProductDialog}>
              <Icon>add_circle</Icon>
            </IconButton>
            <Select
              value={props.newProductId}
              onChange={props.onNewProductSelectChange}
              className={props.classes.selectBox}
            >
              {props.products.filter(product => {
                const found = props.products.find(invoiceProduct => {
                  return invoiceProduct.productId === product._id;
                });
                return !found;
              }).map(product => {
                return (
                  <MenuItem key={product._id} value={product._id}>
                    {product.name}
                  </MenuItem>
                );
              })}
            </Select>
          </ListItem>
          <ListItem style={{width:'20%'}}
            className={props.classes.addFactorProductsNestedListItem}
            key={`ipli_add_2`}
          >
            <TextField
              type="number"
              value={props.newProductCount}
              onChange={props.onNewProductCountChange}
              className="text-center"
            >
            </TextField>
          </ListItem>
          <ListItem style={{width:'20%'}}
            className={props.classes.addFactorProductsNestedListItem}
            key={`ipli_add_3`}
          >
            <Typography>{separateDigits({number: props.newProductUnitPrice, showCurrency: true})}</Typography>
          </ListItem>
          <ListItem style={{width:'20%'}}
            className={props.classes.addFactorProductsNestedListItem}
            key={`ipli_add_4`}
          >
            <Typography>{separateDigits({number: props.newProductTotalPrice})}</Typography>
          </ListItem>
          <ListItem style={{width:'10%'}}
            className={props.classes.addFactorProductsNestedListItem}
            key={`ipli_add_5`}
          >
            <IconButton 
              onClick={onAddNewProductToInvoice}
            >
              <Icon>add_circle</Icon>
            </IconButton>
          </ListItem>
        </List>
      </ListItem>
    </List>
  );
};
