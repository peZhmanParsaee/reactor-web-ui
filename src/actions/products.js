import * as types from './actionTypes';
import * as API from '../api/api';
import * as reduxHelper from '../helpers/reduxHelper';

export const addProduct = (product) => reduxHelper.action(types.ADD_PRODUCT, product);

export const startAddProduct = (productData = {}) => {
  return (dispatch) => {
    const {
      name, 
      stock, 
      unitPrice
    } = productData;

    const product = { name, stock, unitPrice };

    return API.addProduct(product)
      .then((insertedProductRes) => {
        const opStatus = insertedProductRes.data;

        dispatch(addProduct(opStatus.payload));

        return opStatus;
      });
  };
};

export const setProducts = (products) => (
  reduxHelper.action(types.SET_PRODUCTS, products)
);

export const startSetProducts = () => {
  return (dispatch) => {
    return API.getProducts()
      .then((res) => {        
        if (res.data.status === true) {
          dispatch(setProducts(res.data.payload));
        }
      });
  };
};
