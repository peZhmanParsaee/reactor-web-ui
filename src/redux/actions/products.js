import * as types from './actionTypes';
import * as API from '../../api';
import * as reduxHelper from '../../helpers/reduxHelper';

export const addProduct = (product) =>
  reduxHelper.action(types.ADD_PRODUCT, product);

export const startAddProduct = (productData = {}) => async (dispatch) => {
  dispatch(reduxHelper.action(types.ADD_PRODUCT_LOADING));

  const { name, stock, unitPrice } = productData;

  const product = { name, stock, unitPrice };

  const insertedProductRes = await API.addProduct(product);
  const opStatus = insertedProductRes.data;
  dispatch(addProduct(opStatus.payload));

  return opStatus;
};

export const setProducts = (products) =>
  reduxHelper.action(types.SET_PRODUCTS, products);

export const startSetProducts = () => async (dispatch) => {
  const res = await API.getProducts();
  if (res.data.status === true) {
    dispatch(setProducts(res.data.payload));
  }
};
