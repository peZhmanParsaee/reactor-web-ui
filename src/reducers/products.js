import { ADD_PRODUCT, SET_PRODUCTS } from '../actions/types';

const productsReducerDefaultState = [];

export default (state = productsReducerDefaultState, action) => {
  switch(action.type) {
    case ADD_PRODUCT:
      return [
        ...state,
        action.product
      ];
    case SET_PRODUCTS:
      return action.products;
    default:
      return state;
  }
};
