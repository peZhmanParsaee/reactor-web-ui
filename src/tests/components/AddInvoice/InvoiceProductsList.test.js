import React from 'react';
import { shallow } from 'enzyme';
import InvoiceProductsList from '../../../components/AddInvoice/InvoiceProductsList';
import products from '../../fixtures/products';

test('should render InvoiceProductsList with products', () => {
  const wrapper = shallow(
    <InvoiceProductsList products={products} classes={{}} />
  );
  expect(wrapper).toMatchSnapshot();
});
