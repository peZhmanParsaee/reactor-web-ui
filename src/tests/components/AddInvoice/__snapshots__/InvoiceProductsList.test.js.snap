// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render InvoiceProductsList with products 1`] = `
<WithStyles(List)>
  <WithStyles(ListItem)
    key="iplh_1"
  >
    <WithStyles(List)>
      <WithStyles(ListItem)
        key="iplh_1_1"
        style={
          Object {
            "width": "30%",
          }
        }
      >
        نام محصول
      </WithStyles(ListItem)>
      <WithStyles(ListItem)
        key="iplh_1_2"
        style={
          Object {
            "width": "20%",
          }
        }
      >
        تعداد
      </WithStyles(ListItem)>
      <WithStyles(ListItem)
        key="iplh_1_3"
        style={
          Object {
            "width": "20%",
          }
        }
      >
        قیمت واحد
      </WithStyles(ListItem)>
      <WithStyles(ListItem)
        key="iplh_1_4"
        style={
          Object {
            "width": "20%",
          }
        }
      >
        قیمت کل
      </WithStyles(ListItem)>
      <WithStyles(ListItem)
        key="iplh_1_5"
        style={
          Object {
            "width": "10%",
          }
        }
      />
    </WithStyles(List)>
  </WithStyles(ListItem)>
  <WithStyles(InvoiceProductListItem)
    itemKey="ipli_undefined_i"
    key="ipli1_undefined"
    product={
      Object {
        "count": 300,
        "id": undefined,
        "name": "prod 1",
        "totalPrice": undefined,
        "unitPrice": 12100,
      }
    }
  />
  <WithStyles(InvoiceProductListItem)
    itemKey="ipli_undefined_i"
    key="ipli1_undefined"
    product={
      Object {
        "count": 8,
        "id": undefined,
        "name": "prod 2",
        "totalPrice": undefined,
        "unitPrice": 125000,
      }
    }
  />
  <WithStyles(InvoiceProductListItem)
    itemKey="ipli_undefined_i"
    key="ipli1_undefined"
    product={
      Object {
        "count": 652,
        "id": undefined,
        "name": "prod 1",
        "totalPrice": undefined,
        "unitPrice": 14150,
      }
    }
  />
  <WithStyles(ListItem)
    key="ipli_add"
  >
    <WithStyles(List)>
      <WithStyles(ListItem)
        key="ipli_add_1"
        style={
          Object {
            "width": "30%",
          }
        }
      >
        <WithStyles(IconButton)>
          <WithStyles(Icon)>
            add_circle
          </WithStyles(Icon)>
        </WithStyles(IconButton)>
        <WithStyles(WithFormControlContext(Select)) />
      </WithStyles(ListItem)>
      <WithStyles(ListItem)
        key="ipli_add_2"
        style={
          Object {
            "width": "20%",
          }
        }
      >
        <TextField
          className="text-center"
          required={false}
          select={false}
          type="number"
          variant="standard"
        />
      </WithStyles(ListItem)>
      <WithStyles(ListItem)
        key="ipli_add_3"
        style={
          Object {
            "width": "20%",
          }
        }
      >
        <WithStyles(Typography)>
          NaN تومان
        </WithStyles(Typography)>
      </WithStyles(ListItem)>
      <WithStyles(ListItem)
        key="ipli_add_4"
        style={
          Object {
            "width": "20%",
          }
        }
      >
        <WithStyles(Typography)>
          NaN
        </WithStyles(Typography)>
      </WithStyles(ListItem)>
      <WithStyles(ListItem)
        key="ipli_add_5"
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <WithStyles(IconButton)
          onClick={[Function]}
        >
          <WithStyles(Icon)>
            add_circle
          </WithStyles(Icon)>
        </WithStyles(IconButton)>
      </WithStyles(ListItem)>
    </WithStyles(List)>
  </WithStyles(ListItem)>
</WithStyles(List)>
`;
